File: tuan3/hangdoi.cpp
#include <iostream>

const int MAX_SIZE = 100; // Kích thước tối đa của hàng đợi

class Queue {
private:
    int arr[MAX_SIZE];
    int front;  // Chỉ số của phần tử đầu tiên
    int rear;   // Chỉ số của phần tử cuối cùng
    int count;  // Số lượng phần tử hiện tại

public:
    Queue() {
        front = 0;
        rear = -1;
        count = 0;
    }

    // Thêm phần tử vào cuối hàng đợi
    void enqueue(int data) {
        if (isFull()) {
            std::cout << "Hang doi day! Khong the them: " << data << std::endl;
            return;
        }
        rear = (rear + 1) % MAX_SIZE; // Xử lý tràn vòng (circular array)
        arr[rear] = data;
        count++;
    }

    // Xóa và trả về phần tử đầu tiên của hàng đợi
    int dequeue() {
        if (isEmpty()) {
            std::cout << "Hang doi rong! Khong the lay ra." << std::endl;
            return -1; // Hoặc có thể throw exception
        }
        int data = arr[front];
        front = (front + 1) % MAX_SIZE; // Xử lý tràn vòng
        count--;
        return data;
    }

    // Kiểm tra hàng đợi có rỗng không
    bool isEmpty() {
        return count == 0;
    }

    // Kiểm tra hàng đợi có đầy không
    bool isFull() {
        return count == MAX_SIZE;
    }

    // Trả về phần tử đầu tiên của hàng đợi mà không xóa
    int peek() {
      if (isEmpty()){
        std::cout << "Hang doi rong!" << std::endl;
        return -1;
      }
      return arr[front];
    }

    // Trả về kích thước hiện tại của hàng đợi
    int size(){
      return count;
    }
};

int main() {
    Queue q;
    q.enqueue(10);
    q.enqueue(20);
    q.enqueue(30);

    std::cout << "Phan tu dau tien: " << q.peek() << std::endl;

    std::cout << "Lay ra: " << q.dequeue() << std::endl;
    std::cout << "Lay ra: " << q.dequeue() << std::endl;

    q.enqueue(40);
    q.enqueue(50);
    
    std::cout << "Kich thuoc hang doi: "<< q.size() << std::endl;

    while (!q.isEmpty()) {
        std::cout << "Lay ra: " << q.dequeue() << std::endl;
    }

    return 0;
}

File: tuan3/hangdoi.py
from collections import deque

class Queue:
    def __init__(self):
        self.queue = deque()

    # Thêm phần tử vào cuối hàng đợi
    def enqueue(self, item):
        self.queue.append(item)

    # Loại bỏ và trả về phần tử ở đầu hàng đợi
    def dequeue(self):
        if self.is_empty():
            raise IndexError("Dequeue from an empty queue")
        return self.queue.popleft()

    # Kiểm tra xem hàng đợi có rỗng không
    def is_empty(self):
        return len(self.queue) == 0

    # Lấy kích thước của hàng đợi
    def size(self):
        return len(self.queue)

    # Xem phần tử đầu tiên của hàng đợi mà không loại bỏ
    def peek(self):
        if self.is_empty():
            raise IndexError("Peek from an empty queue")
        return self.queue[0]

# Sử dụng hàng đợi
q = Queue()
q.enqueue(10)
q.enqueue(20)
q.enqueue(30)

print("Size of queue:", q.size())  # In ra kích thước hàng đợi
print("Front item:", q.peek())  # In ra phần tử đầu tiên

print("Dequeue:", q.dequeue())  # Xóa và in ra phần tử đầu tiên
print("Size of queue after dequeue:", q.size())  # In ra kích thước sau khi dequeue
